roachprod: make c.Parallel fail fast when a command on any of the nod…

smg260 opened the pull request:
…es fails.

Previously, when encountering an error on a node, we would wait until the command finished on all the specified nodes until returning to the caller. This meant long-running workloads would continue until completion despite having already failed on at least one of the nodes.

The context is now threaded through the call stack, allowing the c.Parallel to issue a cancellation for early termination.

Resolves: #98520
Release note: none
Epic: none

Commit:
roachprod: make c.Parallel fail fast when a command on any of the nodes fails.

Previously, when encountering an error on a node, we would wait until
the command finished on all the specified nodes until returning to the
caller. This meant long-running workloads would continue until completion
despite having already failed on at least one of the nodes.

The context is now threaded through the call stack, allowing the
c.Parallel to issue a cancellation for early termination.

Resolves: #98520
Release note: none
Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/100403)
<!-- Reviewable:end -->

Commit:
roachprod: modify signature of c.Parallel to return ParallelResult,err
as we no longer wait for all the nodes to complete in the event of an
error on any node.

srosenberg reviewed (commented):
> Reviewed 3 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100403)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @renatolabs and @smg260)

___
*[`pkg/roachprod/install/cluster_synced.go` line 2546 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100403#-NSE5rUfELsdVpFXyWQI:-NSE5rUfELsdVpFXyWQJ:b-v4jb2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a33eed15c256bf4cd0d5a7f03dd04050d3bbe815/pkg/roachprod/install/cluster_synced.go#L2546)):*
> ```Go
> 			if r.Err != nil {
> 				groupCancel()
> 				return nil, errors.Wrap(r.Err, "parallel execution failure")
> ```

Looks like the first out-parameter (`ParallelResult`) is no longer needed?

If multiple commands fail concurrently, we'll report whichever one makes it first into the buffered channel; stderr is already captured for each command, so this seems fine?


<!-- Sent from Reviewable.io -->

smg260 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100403)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @renatolabs and @srosenberg)

___
*[`pkg/roachprod/install/cluster_synced.go` line 2546 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100403#-NSE5rUfELsdVpFXyWQI:-NSEX7HM4frGFzmFrj_t:b-kvgc49) ([raw file](https://github.com/cockroachdb/cockroach/blob/a33eed15c256bf4cd0d5a7f03dd04050d3bbe815/pkg/roachprod/install/cluster_synced.go#L2546)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Looks like the first out-parameter (`ParallelResult`) is no longer needed?

If multiple commands fail concurrently, we'll report whichever one makes it first into the buffered channel; stderr is already captured for each command, so this seems fine?
</blockquote></details>

Looks like you viewed an older revision with reviewable (r1 v r2). ParallelResult is returned with the error. 

I wonder why reviewable does that.. 


<!-- Sent from Reviewable.io -->

srosenberg reviewed (approved):
> Reviewed 4 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100403)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @renatolabs and @smg260)

___
*[`pkg/roachprod/install/cluster_synced.go` line 2546 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100403#-NSE5rUfELsdVpFXyWQI:-NSGu3Wp2tD9DEtpXAbL:bpexpq7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a33eed15c256bf4cd0d5a7f03dd04050d3bbe815/pkg/roachprod/install/cluster_synced.go#L2546)):*
<details><summary><i>Previously, smg260 (Miral Gadani) wrote…</i></summary><blockquote>

Looks like you viewed an older revision with reviewable (r1 v r2). ParallelResult is returned with the error. 

I wonder why reviewable does that.. 
</blockquote></details>

You're right; it was indeed r1.


<!-- Sent from Reviewable.io -->

smg260 commented:
TFTR

bors r=srosenberg
